numpy/core/include/numpy/ufuncobject.h

```diff
@@ -209,10 +209,35 @@ typedef struct _tagPyUFuncObject {
          * set by nditer object.
          */
         npy_uint32 iter_flags;
+
+        /* New in version 1 and above */
+
+        /*
+         * for each core_num_dim_ix distinct dimension names,
+         * the possible "frozen" size (-1 if not frozen).
+         */
+        npy_intp *core_dim_sizes;
+
+        /*
+         * for each distinct core dimension, a set of flags OR'd together
+         * e.g., UFUNC_CORE_CAN_IGNORE if signature has ?
+         *       UFUNC_CORE_FROZEN for a numerical dimension
+         */
+        npy_uint32 *core_dim_flags;
+
```

Comment:
The changes need an update to doc/source/reference/c-api.types-and_structures.rst, line 680 or so, with a qualifier that the new fields should only be used if `version > 0`. Come to think of it, we should be checking that as well, just in case someone is mixing headers from old code

Reply:
OK, I see now. From your comments below, you're specifically worried about C code that was compiled with an old include file, and then will try to construct the ufunc with parse_signature but with an old version. Seems like that should be do-able.

Reply:
But it would need a test - would it be enough to just use a new `ufunc` but with version = 0 and then checking the new additions do not get set?

Reply:
Sounds good to me. If you want I can work on this

Reply:
Yes, super! I've spent a bit more time on this than I really should have, and things related to my normal work have been piling up...

Reply:
(Commenting here instead of at https://github.com/mhvk/numpy/pull/1 so that history get kept.)

I'm still a bit confused when exactly this becomes a problem. Suppose I wrote my own ufunc module and have it compiled with the old `ufuncobject.h`, so that internally it counts on the struct not to have `core_dim_flags` and `core_dim_sizes`. Now if I load my module with a new numpy, what exactly happens? From what I saw in cases where I mistakenly loaded astropy in my development branch, it seems one does get a warning automatically, but then code gets run.

Now, presumably, this old module is organized a bit like `_umath_tests.src.c` and thus has a `PyMODINIT_FUNC`, which will run several `PyUFunc_FromFuncAndDataAndSignature`. These in turn will call `parse_signature` and, hey, presto, segmentation fault if that writes to outside the old ufunc struct. So, it does seem one needs a version check in `parse_signature` in addition to having one here that guards against accessing the new additions.

Reply:
hmm. `_parse_signature` is marked `static` so it should never leak out of `PyUFunc_FromFuncAndDataAndSignature`, and the latter allocates its own `PyUFuncObject`. So AFAICT we cannot enter `_parse_signature` with an old struct. My bad, I should have thought this through before requesting changes to this code.

Reply:
Well, that at least is good news!